numpy/core/src/umath/matmul.c.src

```diff
@@ -266,7 +264,45 @@ NPY_NO_EXPORT void
 }
 
 /**end repeat**/
+NPY_NO_EXPORT void
+BOOL_matmul_inner_noblas(void *_ip1, npy_intp is1_m, npy_intp is1_n,
+                           void *_ip2, npy_intp is2_n, npy_intp is2_p,
+                           void *_op, npy_intp os_m, npy_intp os_p,
+                           npy_intp dm, npy_intp dn, npy_intp dp)
+                           
+{
+    npy_intp m, n, p;
+    npy_intp ib1_n, ib2_n, ib2_p, ob_p;
+    char *ip1 = (char *)_ip1, *ip2 = (char *)_ip2, *op = (char *)_op;
+
+    ib1_n = is1_n * dn;
+    ib2_n = is2_n * dn;
+    ib2_p = is2_p * dp;
+    ob_p  = os_p * dp;
 
+    for (m = 0; m < dm; m++) {
+        for (p = 0; p < dp; p++) {
+            *(npy_bool *)op = NPY_FALSE;
+            for (n = 0; n < dn; n++) {
+                npy_bool val1 = (*(npy_bool *)ip1);
+                npy_bool val2 = (*(npy_bool *)ip2);
+                if (val1 != 0 && val2 != 0) {
+                    *(npy_bool *)op = NPY_TRUE;
```

Comment:
in [`BOOL_dot`](https://github.com/numpy/numpy/blob/v1.17.2/numpy/core/src/multiarray/arraytypes.c.src#L3527) there is a break here since we do not need to traverse more of the inner (`dn`) loop once `op` is set to `NPY_TRUE`. However this would require a different style of inner-loop pointer incrementing. Is it worth the effort?

Reply:
Not sure, my guess is that it could might a difference, but boolean matmul is probably super rare. So I do not think it is high priority, and we could wait for someone to complain about the difference in speed. But can't you do:
```
                ip2 += is2_n * (dn - n);
	        ip1 += is1_n * (dn - n);
                break;
```
Which seems OK?

About the tests: My personal opinion is that the regression tests are never the best place to put things. Just the lazy place for bug fixes, and even that is more of a historic thing and we do not use it much anymore.

Reply:
added temporary pointer variables to allow early break out of inner loop.

Reply:
OK, looks good to me. I think it is ready to merge, could try to nitpick, but not worth it ;). To be honest, I think the temp version is maybe clearer anyway.